src/opencl_common.c

```diff
@@ -88,6 +88,55 @@ extern int autotune_get_next_gws_size(size_t num, int step, int startup,
                                       int default_value);
 extern int autotune_get_prev_gws_size(size_t num, int step);
 
+/*
+ * List entry, allocated as (sizeof(struct kernel_opencl) + sizeof(data)).
+ */
+struct kernel_opencl {
+	struct kernel_opencl *next;
+
+	cl_kernel data;
+	cl_kernel *variable;
+};
+
+/*
+ * Main list structure, head is used to start scanning the list, while tail
+ * is used to add new entries.
+ */
+struct opencl_kernels {
+	struct kernel_opencl *head, *tail;
+
+	int count;
+};
+
+/*
+ * List entry, allocated as (sizeof(struct buffer_opencl) + sizeof(data)).
+ */
+struct buffer_opencl {
+	struct buffer_opencl *next;
+
+	cl_mem data;
+	cl_mem *variable;
+	void **host_ptr;
+};
+
+/*
+ * Main list structure, head is used to start scanning the list, while tail
+ * is used to add new entries.
+ */
+struct opencl_buffers {
+	struct buffer_opencl *head, *tail;
+
+	int count;
+};
```

Comment:
I think it'd be convenient for us to start all of these struct names with some prefix indicating they're ours and come from this source file. Like `john_opencl_kernel`, `john_opencl_kernels`, `john_opencl_buffer`, `john_opencl_buffers`. This will become especially important, and will save us the need to rename them later, if we ever have to make them visible to other source files.

Reply:
I've been pushing the use of `opencl_` to indicate this source file. Khronos OpenCL stuff start with `cl_`.

Reply:
OK, if `opencl_` is our convention so far I don't insist on also having `john_`.